### PR TITLE
Add support for selection that spans multiple nodes.

### PR DESCRIPTION
**Pull request by JesusDeLaProg**
Section 3.5.2 of https://wicg.github.io/scroll-to-text-fragment says that while the resulting match can cross the boundaries of a block element, a single fragment item cannot.

For this reason, findText was repurposed to find the deepest block elements that contain an entire fragment item.

In addition to this change, findRangeNodeAndOffset was created to find the exact node and position where the match starts and ends to build the resulting marked section.

Finally, the marked section will be made using Range#surroundContents, which will preserve the DOM structure of the highlighted section. Because we cannot have intersections between DOM Nodes, we need to extend the highlighted section select entire DOM Nodes. For example, if a text fragment ends in the middle of a link, the selection will be extended until the end of the selection is on the same depth than the start of the selection.

Known limitations:

- If the start and the end of a selection are not on the same DOM depth, the selection will be extended until the start and end are on the same depth
- A text fragment cannot cross to another block element, a `<mark>` element will be created and contain both block elements, but because an inline element cannot contain a block element, the text will not be highlighted

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,8 @@
   "description": "This is an attempt to polyfill the [Text Fragments](https://wicg.github.io/scroll-to-text-fragment/) feature for browsers that don't support it natively.",
   "main": "text-fragments.js",
   "dependencies": {
-    "prettier": "^2.0.5"
+    "prettier": "^2.0.5",
+    "terser": "^4.8.0"
```
Comment (tfmar) [inline package.json]: Is this necessary? We run terser as a plugin on rollup.

File: src/text-fragment-utils.js [hunk 2]
```diff
@@ -16,7 +16,7 @@
 
 /**
  * @typedef {Object} TextFragment
- * @property {string} textStart
+ * @property {string?} textStart
```
Comment (tfmar) [inline src/text-fragment-utils.js:19]: I'm pretty sure the spec requires textStart; when would this be missing?

File: text-fragment-utils.js [hunk 3]
```diff
@@ -83,79 +143,171 @@ const escapeRegExp = (s) => {
...
   }
+  if (mark.parentElement) {
+    return mark;
+  } else {
+    return;
+  }
+};
+
+/**
+ * Scrolls an element into view, folloring the recommandation of
```
Comment (tfmar) [inline text-fragment-utils.js]: spelling nit: following, recommendation

File: src/text-fragment-utils.js (hunk 2 above)
Comment (tomayac) [reply src/text-fragment-utils.js:19]: +1, textStart is the one required parameter.

**Comment by tomayac** [comment]
> For example, if a text fragment ends in the middle of a link, the selection will be extended until the end of the selection is on the same depth than the start of the selection.

Not sure I fully understand what you mean here. Selections can cover only part of a link, see https://example.com/#:~:text=More.

**Comment by tomayac** [comment]
The polyfill is in a _very_ early stage right now, and while I appreciate the PR, I'm inclined to closing it for the time being. Google has dedicated engineering resources (👋 @tfmar) to develop the polyfill and we have specific plans for it that have certain not yet public requirements for how the code needs to be structured. @tfmar, what do you think?

**Comment by tfmar** [comment]
> > For example, if a text fragment ends in the middle of a link, the selection will be extended until the end of the selection is on the same depth than the start of the selection.
> 
> Not sure I fully understand what you mean here. Selections can cover only part of a link, see https://example.com/#:~:text=More.

I think the issue described here is that if the page is: `<p> This is some <a href="example.com">highlighted text</a></p>`

then highlighting the phrase "some highlighted" would produce `<mark> some <a> highlighted </mark> text </a>` which isn't valid HTML.

**Comment by tomayac** [comment]
> > > For example, if a text fragment ends in the middle of a link, the selection will be extended until the end of the selection is on the same depth than the start of the selection.
> > 
> > 
> > Not sure I fully understand what you mean here. Selections can cover only part of a link, see https://example.com/#:~:text=More.
> 
> I think the issue described here is that if the page is: `<p> This is some <a href="example.com">highlighted text</a></p>`
> 
> then highlighting the phrase "some highlighted" would produce `<mark> some <a> highlighted </mark> text </a>` which isn't valid HTML.

Oh, gotcha. Yeah, modifying HTML with RegExp can only go wrong :-) We need to come up with a smarter way of doing this in the future. It worked well enough for the very first proof of concept. 

I also noticed that having two text fragments in the same paragraph is destructive…

**Comment by JesusDeLaProg** [comment]
> Oh, gotcha. Yeah, modifying HTML with RegExp can only go wrong :-) We need to come up with a smarter way of doing this in the future. It worked well enough for the very first proof of concept.
> 
> I also noticed that having two text fragments in the same paragraph is destructive…

Yes, both those problems is why I went with the Range approach instead of RegExp, because it preserves the HTML structure. Right now, you can have as many (non-intersecting) text fragment in the same block and it won't break the HTML structure or remove content from the page.

Another limitation that should be noted is that you cannot have multiple block elements in your selection, because `<mark>`, an inline element, cannot contain a block element. Example:
```html
<ul>
<mark>
<li>One highlighted item</li>
<li>Another highlighted item</li>
</mark>
</ul>
```
The resulting HTML will be accepted by the UA, but the content will not be highlighted, `<mark>` will effectively be ignored.

Even with these limitation, it is still something that works and could be worked on further.

File: src/text-fragment-utils.js (hunk 2 above)
Comment (JesusDeLaProg) [reply src/text-fragment-utils.js:19]: Sorry, I just double checked and the properties are never null, they are either empty strings or strings with content, removing the optional mark.

File: package.json (hunk 1 above)
Comment (JesusDeLaProg) [reply package.json]: I didn't see rollup-plugin-terser, removing this because it's not necessary.

File: dist/text-fragments.js [hunk 4]
```diff
@@ -1 +0,0 @@
...&c[0].parentNode===r[0].parentNode){const t=r[0].parentNode,o=r[0].textContent.replace(new RegExp(`(^.*?)(${n(e.textStart)})(.*?)(${e.textEnd})(.*?$)`,"m"),"$1<mark>$2$3$4</mark>$5");t.textContent="",t.insertAdjacentHTML("afterbegin",o),t.scrollIntoView(i)}}else{const t=r[0].parentNode,o=r[0].textContent.replace(new RegExp(`(^.*?)(${n(e.textStart)})(.*?$)`,"m"),"$1<mark>$2</mark>$3");t.textContent="",t.insertAdjacentHTML("afterbegin",o),t.scrollIntoView(i)}t.length},o=e=>{if(!e)return[];const t=document.body,n=document.createTreeWalker(t,NodeFilter.SHOW_TEXT,{acceptNode:t=>{if(t.textContent.includes(e))return NodeFilter.FILTER_ACCEPT}}),r=[];let o=n.currentNode;for(;o;)r.push(o),o=n.nextNode();return r.slice(1)};(async()=>{const n=document.location.hash;if("fragmentDirective"in Location.prototype||!n)return;Location.prototype.fragmentDirective={};(e=>{const t={};for(const[n,o]of Object.entries(e))"text"===n&&(t[n]=o.map(e=>r(e)))})((e=>{const n={};for(const[r,o]of Object.entries(e))"text"===r&&(n[r]=o.map(e=>t(e)));return n})((t=>{const n=t.replace(/#.*?:~:(.*?)/,"$1");if(!n)return;const r=new URLSearchParams(n),o={};return e.forEach(e=>{r.has(e)&&(o[e]=r.getAll(e))}),o})(n)))})();
```
Comment (tfmar) [inline dist/text-fragments.js:1]: What's the rationale behind deleting this? I know anyone who installs the package will have it generated for them, but I also think having a copy ready to go makes it easier to test/integrate without having to install the package (e.g., if your project doesn't use npm). 

I guess the risk is that we forget to update this before commits?

Comment (tfmar) [reply dist/text-fragments.js:1]: We discussed this offline and decided it's best to leave this unpopulated.

Comment (tomayac) [reply dist/text-fragments.js:1]: This post summarized the rationale quite well: https://hackernoon.com/how-to-publish-your-package-on-npm-7fc1f5aae600. There’s https://unpkg.com/text-fragments-polyfill for quickly getting the /dist folder, or just npm install.

**Pull request merged**